File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.py
from collections import deque
from sys import stdin; input = stdin.readline

def bfs(start):
    queue = deque([start])

    visited[start] = True

    while queue:
        v = queue.popleft()

        for i in graph[v]:
            if not visited[i]:
                queue.append(i)
                visited[i] = True

N,M = map(int,input().split())

visited = [False] * (N+1)

graph = [[] for _ in range(N+1)]

ans = 0

for _ in range(M):
    u,v = map(int,input().split())
    graph[u].append(v)
    graph[v].append(u)

for i in range(1, N+1):
    if not visited[i]:
        bfs(i)
        ans+=1
print(ans)

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.js
const fs = require("fs");

const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

function solution(arr) {
  function bfs(start) {
    visited[start] = true;
    const queue = [start];
    while (queue.length !== 0) {
      const v = queue.shift();
      graph[v].forEach((el) => {
        if (!visited[el]) {
          visited[el] = true;
          queue.push(el);
        }
      });
    }
  }

  const [N, M] = arr[0].split(" ").map(Number);

  const graph = Array.from({ length: N + 1 }, () => []);
  const visited = Array.from({ length: N + 1 }, () => false);

  let ans = 0;

  for (let i = 0; i < M; i++) {
    const [u, v] = input[i + 1].split(" ").map(Number);
    graph[u].push(v);
    graph[v].push(u);
  }

  for (let i = 1; i <= N; i++) {
    if (!visited[i]) {
      bfs(i);
      ans++;
    }
  }
  return ans;
}

console.log(solution(input));
